notebooks/metatranscriptomics_pipeline_iSeqnorm.ipynb

```diff
@@ -0,0 +1,1477 @@
+   "source": [
+    "### gDNA concentration heatmap, Plate 4"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "plot_plate_vals(dna_concs[np.ix_(odd_rows,odd_cols)],\n",
+    "                    annot_str= names[np.ix_(odd_rows,odd_cols)],\n",
+    "                    color_map='viridis',\n",
+    "                    annot_fmt='')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Workflow for assigning barcodes\n",
+    "\n",
+    "This portion of the notebook will assign index values and construct an Echo picklist file for adding barcodes. \n",
+    "\n",
+    "As inputs, it requires:\n",
+    "1. A plate map dataframe (from previous step)\n",
+    "2. A tab-delimited index combination file, relating index combinations, i5 and i7 index values, and i5 and i7 index locations\n",
+    "\n",
+    "The workflow then:\n",
+    "1. reads in the index combo list\n",
+    "2. assigns indices per sample\n",
+    "3. produces an Echo-formatted pick list file"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Step 1: Read in index combo list\n",
+    "\n",
+    "This is a file that contains every possible i5 and i7 barcode combo on a separate line,\n",
+    "along with plate and well location information. It should look something like this:\n",
+    "\n",
+    "```\n",
+    "index combo,index combo seq,i5 name,i5 sequence,i5 well,i5 plate,i7 name,i7 sequence,i7 well,i7 plate\n",
+    "0,ACCGACAAACGTTACC,iTru5_01_A,ACCGACAA,A1,iTru5_plate,iTru7_101_01,ACGTTACC,A1,iTru7_plate\n",
+    "1,AGTGGCAACTGTGTTG,iTru5_01_B,AGTGGCAA,B1,iTru5_plate,iTru7_101_02,CTGTGTTG,A2,iTru7_plate\n",
+    "2,CACAGACTTGAGGTGT,iTru5_01_C,CACAGACT,C1,iTru5_plate,iTru7_101_03,TGAGGTGT,A3,iTru7_plate\n",
+    "3,CGACACTTGATCCATG,iTru5_01_D,CGACACTT,D1,iTru5_plate,iTru7_101_04,GATCCATG,A4,iTru7_plate\n",
+    "4,GACTTGTGGCCTATCA,iTru5_01_E,GACTTGTG,E1,iTru5_plate,iTru7_101_05,GCCTATCA,A5,iTru7_plate\n",
+    "5,GTGAGACTAACAACCG,iTru5_01_F,GTGAGACT,F1,iTru5_plate,iTru7_101_06,AACAACCG,A6,iTru7_plate\n",
+    "6,GTTCCATGACTCGTTG,iTru5_01_G,GTTCCATG,G1,iTru5_plate,iTru7_101_07,ACTCGTTG,A7,iTru7_plate\n",
+    "7,TAGCTGAGCCTATGGT,iTru5_01_H,TAGCTGAG,H1,iTru5_plate,iTru7_101_08,CCTATGGT,A8,iTru7_plate\n",
+    "8,CTTCGCAATGTACACC,iTru5_02_A,CTTCGCAA,I1,iTru5_plate,iTru7_101_09,TGTACACC,A9,iTru7_plate\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "index_combo_fp = './test_output/iTru/new_iTru_combos_Dec2017.csv'\n",
+    "\n",
+    "if not os.path.isfile(index_combo_fp):\n",
+    "    print(\"Problem! %s is not a path to a valid file\" % file)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "index_combos = pd.read_csv(index_combo_fp)\n",
```

Comment:
Daniel deeply wants read_csv() calls to use dtype=str so that they read everything in as a string rather than allowing automatic type selection (which profoundly screws up things like tube codes).  However, my guess is that you might want the index column to be read in as an int.  Would you consider reading everything in as a str and then explicitly converting the columns you need to be not strings into the type you want them to be?

Reply:
It is interesting that historically we've wrapped sample-sheets in objects that incorporate custom type conversion and validation, but we don't do it for csvs of other types. At the least we should migrate people towards using a wrapper function that loads csvs into data frames as all dtype=str. this would make periodically searching and replacing new instances of read_csv() a non-issue. custom openers for different csvs with different headers might be a possible solution as well.

Reply:
I am open to making the change, but do we want to do that during this pull request?

Reply:
Yeah, I honestly don't know. What I learned from making changes to the matrix notebook is that it is not safe to just add a `dtype=str` to an existing `read_csv`, because in some places, later functionality depends on some columns being, e.g., numbers.  You have to actually know what is being read in and how it is being used later before it is safe to make this change.  I was thinking Rodolfo might have the knowledge necessary to make the change safely here, but if not, I don't want to slow things down for it.

I do not think I am up for trying to add a wrapper function in this PR since we need it to be merged as soon as we can.

Reply:
I agree it doesn't sound like the time and place for a wrapper. This comment doesn't appear to be outdated. Can I assume it was addressed?

Reply:
I can live with leaving this out for now.